### DISCUSSION TITLE
Inline SVG for sprites

### DISCUSSION THREAD

**Discussion by arnaudroques** [plantuml/plantuml#1066] [General]
[Last snapshot](https://github.com/plantuml/plantuml/releases/tag/snapshot) has an undocumented feature. You can now use inlined SVG for sprites.

For example:

```
@startuml
sprite foo1 <svg viewBox="0 0 36 36">
<path fill="#77B255" d="M36 32c0 2.209-1.791 4-4 4H4c-2.209 0-4-1.791-4-4V4c0-2.209 1.791-4 4-4h28c2.209 0 4 1.791 4 4v28z"/>
<path fill="#FFF" d="M21.529 18.006l8.238-8.238c.977-.976.977-2.559 0-3.535-.977-.977-2.559-.977-3.535 0l-8.238 8.238-8.238-8.238c-.976-.977-2.56-.977-3.535 0-.977.976-.977 2.559 0 3.535l8.238 8.238-8.258 8.258c-.977.977-.977 2.559 0 3.535.488.488 1.128.732 1.768.732s1.28-.244 1.768-.732l8.258-8.259 8.238 8.238c.488.488 1.128.732 1.768.732s1.279-.244 1.768-.732c.977-.977.977-2.559 0-3.535l-8.24-8.237z"/>
</svg>

Alice->Bob : hello <$foo1> there

@enduml
```

or

```
@startuml
sprite foo1 <svg width="8" height="8" viewBox="0 0 8 8"><path d="M1 0l-1 1 1.5 1.5-1.5 1.5h4v-4l-1.5 1.5-1.5-1.5zm3 4v4l1.5-1.5 1.5 1.5 1-1-1.5-1.5 1.5-1.5h-4z" /></svg>

Alice->Bob : hello <$foo1> there

@enduml
```

The implementation is done through [simple classes](https://github.com/plantuml/plantuml/blob/master/src/net/sourceforge/plantuml/emoji/SvgNanoParser.java), so only few SVG directives are managed (`path`, `ellipse`, `circle`...). This means that you will probably have to compress your SVG (with something [like  this](https://vecta.io/nano)).

However, the good news is that even if your sprite is SVG, diagrams could be generated to PNG/SVG/EPS (because SVG is parsed by PlantUML).

There are likely bugs here, so please tell us if you find any. The big question is also to know whether this is useful or not.

Thanks!

**Comment by The-Lum** [comment]
Hello @arnaudroques, PlantUML team, and all,

Very good... 👍 
>The big question is also to know whether this is useful or not.

Yes it is very useful especially to have a **_complete SVG stack_**, and a PNG output as a fallback.
Then the big deal... will be how to migrate all [`stdlib`](https://github.com/plantuml/plantuml-stdlib) to profit SVG power...


_Have a good summer. 🌞_
Regards.

**Comment by Potherca** [comment]
Currently, I help maintain several icon repositories (under https://github.com/plantuml-stdlib/). Most of these libs add icons/sprites to PlantUML (through https://github.com/plantuml/plantuml-stdlib/), by converting SVG to PNG to PlantUML Sprite.

Each of these projects has their own conversion logic. Because of little quirks in the SVGs, the conversion tools used and/or PlantUML, a lot of time is spent/wasted this logic.


So, regarding:

> The big question is also to know whether this is useful or not.

Being able to create sprites from SVG directly would be **HUGE** for these projects!

It would mean a lot of sprite generation logic could be replaced by much simpler logic or simply removed. That would make it possible to spend time or automating the updates from source material, rather than having to maintain generation logic.

**Reply by npiper** [reply]
Perhaps watch for progress on these that have gone in for plantuml to use in your plantuml-stdlib repo:
  - this could also improve the rendering in a lot of these great libraries you have created and maintained.

#2548

#2506

**Reply by npiper** [reply]
@Potherca this has now been merged to main if you'd like to try in a beta version, needs a `pragma` choice to force it.

#2568

**Comment by hakanson** [comment]
Any plans on supporting `<text>` element?  I was experimenting with numbered callouts using a black circle with white text.  The text doesn't show up and I tested the SVG in my web browser.

```
@startuml

sprite callout1 <svg xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 18 18">
  <circle cx="9" cy="9" r="9" fill="black" />
  <text x="9" y="10" stroke="white" fill="white" stroke-width="1px" text-anchor="middle" alignment-baseline="middle">1</text>
</svg>

Alice->Bob: <$callout1> message

@enduml
```

![image](https://user-images.githubusercontent.com/414500/188509603-2888e575-1bc6-481e-837c-362c01d99800.png)

**Reply by arnaudroques** [reply]
> Any plans on supporting `<text>` element? I was experimenting with numbered callouts using a black circle with white text. The text doesn't show up and I tested the SVG in my web browser.

Nice idea!
We would like to keep our internal SVG parser simple and light so we won't support all <text> attributes.
However, [with last snapshot](https://github.com/plantuml/plantuml/releases/tag/snapshot), you can now have:

```
@startuml
sprite callout1 <svg xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 18 18">
  <circle cx="9" cy="9" r="9" fill="black" />
  <text x="6" y="11" fill="yellow" font-size="10">1</text>
</svg>

Alice->Bob: <$callout1> message
@enduml
```

Hope this helps!

**Comment by marcosngomezi** [comment]
Hi there its a really useful feature, thanks for that, Im having problems with rotate transform, looks like its just rotating the center position, also looks like its not responding to the axis of rotation coordinates.
The SVG its a simple react logo:
`sprite react <svg viewBox="0 0 230 230">
<circle cx="115" cy="102" r="20.5" fill="#61dafb"/>
<ellipse rx="110" ry="42"  cx="115" cy="102" stroke="#61dafb" stroke-width="10" fill="none"/>
<ellipse rx="110" ry="42"  cx="115" cy="102" transform="rotate(60 115 102)" stroke="#61dafb" stroke-width="10" fill="none"/>
<ellipse rx="110" ry="42"  cx="115" cy="102" transform="rotate(-60 115 102)" stroke="#61dafb" stroke-width="10" fill="none"/>
</svg>
`
[test](https://www.plantuml.com/plantuml/uml/lP6nJiGm38RtFeLb660ucXQugK4dtYIaRed89IVRz428TyTNWLXWX25APFlxyalUYGRMvtrk6ZbmKWAc41K6MMPO4Xtlwuj71mxww_NWNJF4n34JHEjqtIr2VBM7wn7OO--kh3Abd3sURRinJ0_O6aCvfuEOVeMSpHfqOymF6b6kJ_JDVnOsnpJgusdcwuDI2wts6PH3aQdotYDN3KeNMmUc1JDUufy7QETVKcp-FqOhYrqDKzHGvammdAzBVPCOCldktPexAkDfwny0)

Also it breaks when rotation is higher to 89 or lower to -89. For my case its not a problem, but maybe for other images it is.
Thanks for the work!

**Reply by arnaudroques** [reply]
Many thanks for the feedback

This should be fixed [with last snapshot](https://github.com/plantuml/plantuml/releases/tag/snapshot).

We probably also fix the issue with larger rotation ([test](https://www.plantuml.com/plantuml/uml/hP7FIiGm48VlVOeXUj33sadzSsgMnJU9wQG6iyao6RkA-EwEXRqfBE90QPZCxyi7islYMLxtkMi7JaB0v8F0s9OPbaJ7n_hc30B2SBj-PjkD8N78145lhBqt4DxrW8C1TcR06-t4bBCp5mzsyl79z9gXdDEXANuDeSvgw4upFs2QS7qXpSU8MgV6vfWcUVwUEJrGQg6LFeEmBorMtZl3LRpGbKK4vO8YhyrV3H1ZFCkWdt_HsFoFXrQCvthqRT5VnnJ4bpaJZ9VhMZzQy9cS_TnsEogJRlqB)).
Could you tell us if it's ok for you?

Note that there _must_ be other issues, so do not hesitate to post again!

Thanks

**Reply by marcosngomezi** [reply]
Nice!, works like a charm now, really thanks for the fast response and fix.

**Reply by marcosngomezi** [reply]
Im back lol, looks like is not reacting nice to scaling the sprite. 
https://www.plantuml.com/plantuml/uml/lP71IWCn48RlUOfXzA07soHgJvkb-2OnErk3QL8cOwk8x-wmL0JngeU18Pdl-m__jajaUTaNqm-SXO0f9e6X7tTmp7HwQAy1BLZmjylWQ8QKEHM2f3_EtIEaj_F20RrTwJBdKW9URDmKvqTSAqEbv4DN_G9P15RepYFyekd2xPc--VF3pIbFygGtzYkWjaf_R0VXMFlSU1-GcqIXguq55O7EzV-7E_yZVQrDZSOm9ObrLmY6owN2zvvYeM1N_cCqMwgJTlm9

**Reply by arnaudroques** [reply]
Thanks for the feedback!
[This is fixed](https://www.plantuml.com/plantuml/uml/lP6nJWCn38RtF8Kn333GvW9qkfmgtYJaV2KYJIh7j4M8TyUSA0DYfOCbA_7t_SE_RX9ONhVPI99CCAKZZHUcxJW90LE80dtRRs2VwF1OZnujM71ty-1W-fWuAXVrf-iU4EBRzy8UdLteCgMSFLwkkZ5CJxXKXd9EkwRw6R88hD2zG_X3quJh2_tmfuVRGnhbMM_iAQ3KGlzi1-5GsbHvwv6h1A7hbGKLVStD-SCxzojzgKqDnZ152MMJ2VghkSBt5aCcRnVkOp1hAgCM_Wa0).
Waiting for next issue :-)

**Comment by npiper** [comment]
Would be great to see this progress further - could an SVG library be 'plugged in' to PlantUML or are we talking significant surgery for Sprites to do this?  New Object type?

Some SVG libraries;
https://xmlgraphics.apache.org/batik/
https://github.com/jfree/jfreesvg

**Comment by Palkers76** [comment]
This would be very usefull
I have planty of svg icons which are not in standard libraries now. Some of them are corporate one So cant be public
Do you think it will be released some day?

**Comment by npiper** [comment]
Created a Feature / Pull Request to add an alternative SVG Parsing approach (DOM Batik) -

 it could be useful to test if Library creations or stdlib libraries with more extensive SVG element support could use inline SVG sprite definitions with the approach vs. converting using the encoder or sprite import utils available.
 

https://github.com/plantuml/plantuml/pull/2497

**Reply by npiper** [reply]
Updated to a more minimal SAX based parser in this Issue, which now has a PR

#2548 #2506
